sticky-rows/StickyRows.js

```diff
@@ -0,0 +1,186 @@
+
+    const spreadsheet = SpreadsheetApp.getActiveSpreadsheet();
+    if (!spreadsheet) {
+        throw new Error("Must run in the context of a Spreadsheet");
+    }
+
+    const sheet = spreadsheet.getActiveRange().getSheet();
+
+    // setup config for this sheet in the spreadsheet attributes
+    const key = "stickyRows." + sheet.getSheetId();
+    const config = JSON.parse(PropertiesService.getDocumentProperties().getProperty(key) || '{}');
+    if (config.dynamic_range !== dynamic_range || config.sticky_range !== sticky_range) {
+        config.dynamic_range = dynamic_range;
+        config.sticky_range = sticky_range;
+        PropertiesService.getDocumentProperties().setProperty(key, JSON.stringify(config));
+    }
+
+    // ensure onChange() trigger
+    let haveTrigger = false;
+    for (const trigger of ScriptApp.getProjectTriggers()) {
+        if (trigger.getEventType() === ScriptApp.EventType.ON_CHANGE && trigger.getHandlerFunction() === TRIGGER_FUNCTION_NAME) {
+            haveTrigger = true;
+        }
+    }
+
+    if (!haveTrigger) {
+        throw new Error(`Please install function ${TRIGGER_FUNCTION_NAME}() as trigger in this Spreadsheet's script context (event source: From spreadsheet, event type: onChange)`)
+    }
+}
+
+
+/** Uninstall triggers. */
+function uninstall() {
+    TriggerUtil.uninstall(TRIGGER_HANDLER_FUNCTION);
+}
+
+
+/** Install spreadsheet onChange trigger. */
+function install() {
+
+    TriggerUtil.uninstall(TRIGGER_HANDLER_FUNCTION);
+
+    Logger.log("Installing onChange trigger for spreadsheet %s", spreadsheet.getId());
+
+    const sheet = SpreadsheetApp.getActive();
+    ScriptApp.newTrigger(TRIGGER_HANDLER_FUNCTION)
+        .forSpreadsheet(SpreadsheetApp.getActiveSpreadsheet().getId())
+        .onChange()
+        .create();
+    Logger.log("Installed onChange trigger for spreadsheet %s", sheet.getSpreadsheet());
+}
+
+
+/** Perform the actual sticky rows logic.
+ *
+ * Supposed to be called in a SpreadSheet.onChange() trigger.
+ */
+function onChangeStickyRows(event) {
+
+    const spreadsheet = event.source;
+    if (!spreadsheet) {
+        throw new Error("Must run in the context of a Spreadsheet");
+    }
+
+    const sheet = spreadsheet.getActiveRange().getSheet();
+    const key = "stickyRows." + sheet.getSheetId();
+    const config = JSON.parse(PropertiesService.getDocumentProperties().getProperty(key) || '{}');
+    if (!config.dynamic_range || !config.sticky_range) {
+        throw new Error("dynamic_range and sticky_range must be set");
+    }
+
+    const dynamic = sheet.getRange(config.dynamic_range);
+    const sticky = sheet.getRange(config.sticky_range);
+
+    const rowCount = sticky.getNumRows();
+    if (dynamic.getNumRows() !== rowCount) {
+        throw new Error("dynamic/sticky ranges must have the same number of rows");
+    }
+
+    if (dynamic.getColumn() < sticky.Right && dynamic.getColumn() + dynamic.getNumColumns() > sticky.getColumn() &&
```

Comment:
Forwarding the errors in a meaningful/reliable way was unexpectedly **hard**.

There are a lot of mechanisms at work that prevent it, to list a few challenges:
- Google Sheets analyzes formulas and ignores white-space changes in API setFormula() calls
- Formulas can have different argument separators (;,) depending on locale
- Boolean values are represented by different words based on locale (TRUE, WAHR, ...)
- There are several formula parsing libraries for Google Sheets/Excel out there, but they'd be 100% overkill: https://github.com/formulajs/formulajs/blob/master/src/utils/common.js
- There seems to be additional logic breaking `cell change -> trigger -> cell change` execution cycles for many types of changes that the UI will trigger re-calculates for (if a human does it).
- Since we are overwriting the formula, we must take great care to only do so when applicable (ensure control cell is set and valid).

Reply:
A tested and reliable way to solve this is adding a `toggle flag` or incrementing number as an argument to the `STICKY_ROWS()` function.